sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -384,17 +385,31 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
     val bis = new ByteArrayInputStream(bytes)
     val ins = new DataInputStream(codec.compressedInputStream(bis))
 
-    new Iterator[InternalRow] {
+    new NextIterator[InternalRow] {
+      Option(TaskContext.get()).foreach(_.addTaskCompletionListener[Unit](_ => closeIfNeeded()))
       private var sizeOfNextRow = ins.readInt()
```

Comment:
nit: should we also protect this with `try .. catch` and close the input when error happens?

Reply:
I'm wondering if `addTaskCompletionListener` is actually necessary. Isn't it the case that `TaskContext` is only available on executor threads?

Reply:
Yea seems you are right. `decodeUnsafeRows` is only called on collected result from driver. Hmm in this case I wonder if we should find other ways to make sure the input stream is closed at the end.

Reply:
I think one problem is that `Dataset.toLocalIterator` indirectly uses `decodeUnsafeRows`:
```scala
Dataset:
  def toLocalIterator(): java.util.Iterator[T] = {
    withAction("toLocalIterator", queryExecution) { plan =>
      val fromRow = resolvedEnc.createDeserializer()
      plan.executeToIterator().map(fromRow).asJava
    }
  }

SparkPlan:
  def executeToIterator(): Iterator[InternalRow] = {
    getByteArrayRdd().map(_._2).toLocalIterator.flatMap(decodeUnsafeRows)
  }
```
since the iterator is transferred to clients after `Dataset.toLocalIterator`, there's no way for Spark to know how the iterator will be used, and whether it will be completely drained. Therefore, it seems impossible to know when we should call `close` on the input stream.

Reply:
Maybe we can use `CompletionIterator` to make sure `close` is called when this iterator reaches the end?

Reply:
Oh, based on https://github.com/apache/spark/pull/35613#discussion_r812633201, when `getNext` reaches the end, `close` will be called. I think we may not need to set another way to call it.

Reply:
Both `NextIterator` and `CompletionIterator` suffer from the same limitation, ie. if the consumer of the Iterator does not consume the Iterator in its entirety, either because of an exception when processing an item or when using a method like `Iterator.find` for example, then neither will call `close`.

Reply:
But I would agree that `CompletionIterator` provides a neater solution.

Reply:
Mmmm, it seems about the same:

```
    val iter: Iterator[InternalRow] = new Iterator[InternalRow] {
      private var sizeOfNextRow = ins.readInt()
      override def hasNext(): Boolean = sizeOfNextRow >= 0
      override def next(): InternalRow = {
        try {
          val bs = new Array[Byte](sizeOfNextRow)
          ins.readFully(bs)
          val row = new UnsafeRow(nFields)
          row.pointTo(bs, sizeOfNextRow)
          sizeOfNextRow = ins.readInt()
          row
        } catch {
          case t: Throwable if ins != null =>
            ins.close()
            throw t
        }
      }
    }
    
    new CompletionIterator(iter) {
      def completion(): Unit = ins.close()
    }
```